src/Core/Waiter.php

```diff
@@ -0,0 +1,222 @@
+<?php
+
+declare(strict_types=1);
+
+namespace AsyncAws\Core;
+
+use AsyncAws\Core\Exception\Http\HttpException;
+use AsyncAws\Core\Exception\Http\NetworkException;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+/**
+ * The result promise is always returned from every API call. Remember to call `resolve()` to
+ * make sure the request is actually sent.
+ */
+class Waiter
+{
+    public const STATE_SUCCESS = 'success';
+    public const STATE_FAILURE = 'failure';
+    public const STATE_PENDING = 'pending';
+
+    protected const WAIT_TIMEOUT = 30.0;
+    protected const WAIT_DELAY = 5.0;
+
+    /**
+     * @var AbstractApi|null
+     */
+    protected $awsClient;
+
+    /**
+     * @var array|object|null
+     */
+    protected $request;
+
+    /**
+     * @var ResponseInterface|null
+     */
+    private $response;
+
+    /**
+     * @var HttpClientInterface
+     */
+    private $httpClient;
+
+    /**
+     * @var string|null
+     */
+    private $finalState;
+
+    public function __construct(ResponseInterface $response, HttpClientInterface $httpClient, AbstractApi $awsClient, $request)
+    {
+        $this->response = $response;
+        $this->httpClient = $httpClient;
+        $this->awsClient = $awsClient;
+        $this->request = $request;
+    }
+
+    public function __destruct()
+    {
+        if (isset($this->response)) {
+            $this->response->cancel();
+        }
+    }
+
+    final public function isSuccess(): bool
+    {
+        return self::STATE_SUCCESS === $this->getState();
+    }
+
+    final public function isFailure(): bool
+    {
+        return self::STATE_FAILURE === $this->getState();
+    }
+
+    final public function isPending(): bool
+    {
+        return self::STATE_PENDING === $this->getState();
+    }
+
+    final public function getState(): string
+    {
+        if (null !== $this->finalState) {
+            return $this->finalState;
+        }
+
+        if (!isset($this->response)) {
+            $this->stealResponse($this->refreshState());
+        }
+
+        $exception = null;
+        $this->resolve();
+
+        \assert(null !== $this->response);
```

Comment:
We never unset the response in Resolve(). $this->response will always have a value, right?

Reply:
yes. your right.

But psalm yield because `$this->response` can be null. I hesitate between ignoring the psalm error and adding this `assert`.

Reply:
I think psalm complained because the property is nullable.

Reply:
yes exactly. The property is nullable, and psalm yield. 
But as you sayed, in the current workflow, nothing unset the property and it can't be null.

Should I ignore the error reported by psalm? Or add a useless `assert`?

Reply:
hmm. removed and nobody complains anymore... weird.